subsys/disk/disk_access_stm32_sdmmc.c

```diff
@@ -0,0 +1,420 @@
+		priv->status = DISK_STATUS_UNINIT;
+	} else {
+		LOG_DBG("card removed");
+		stm32_sdmmc_access_deinit(priv);
+		priv->status = DISK_STATUS_NOMEDIA;
+	}
+}
+
+static void stm32_sdmmc_cd_callback(struct device *gpiodev,
+				    struct gpio_callback *cb,
+				    u32_t pin)
+{
+	struct stm32_sdmmc_priv *priv = CONTAINER_OF(cb,
+						     struct stm32_sdmmc_priv,
+						     cd_cb);
+
+	k_work_submit(&priv->work);
+}
+
+static int stm32_sdmmc_card_detect_init(struct stm32_sdmmc_priv *priv)
+{
+	int err;
+
+	if (!priv->cd.name) {
+		return 0;
+	}
+
+	priv->cd.port = device_get_binding(priv->cd.name);
+	if (!priv->cd.port) {
+		return -ENODEV;
+	}
+
+	gpio_init_callback(&priv->cd_cb, stm32_sdmmc_cd_callback,
+			   1 << priv->cd.pin);
+
+	err = gpio_add_callback(priv->cd.port, &priv->cd_cb);
+	if (err) {
+		return err;
+	}
+
+	err = gpio_pin_configure(priv->cd.port, priv->cd.pin,
+				 priv->cd.flags | GPIO_INPUT);
+	if (err) {
+		goto remove_callback;
+	}
+
+	err = gpio_pin_interrupt_configure(priv->cd.port, priv->cd.pin,
+					   GPIO_INT_EDGE_BOTH);
+	if (err) {
+		goto unconfigure_pin;
+	}
+	return 0;
+
+unconfigure_pin:
+	gpio_pin_configure(priv->cd.port, priv->cd.pin, GPIO_DISCONNECTED);
+remove_callback:
+	gpio_remove_callback(priv->cd.port, &priv->cd_cb);
+	return err;
+}
+
+static int stm32_sdmmc_card_detect_uninit(struct stm32_sdmmc_priv *priv)
+{
+	if (!priv->cd.name) {
+		return 0;
+	}
+
+	gpio_pin_interrupt_configure(priv->cd.port, priv->cd.pin,
+				     GPIO_INT_MODE_DISABLED);
+	gpio_pin_configure(priv->cd.port, priv->cd.pin, GPIO_DISCONNECTED);
+	gpio_remove_callback(priv->cd.port, &priv->cd_cb);
+	return 0;
+}
+
+static int stm32_sdmmc_pwr_init(struct stm32_sdmmc_priv *priv)
+{
+	int err;
+
+	if (!priv->pe.name) {
+		return 0;
+	}
+
+	priv->pe.port = device_get_binding(priv->pe.name);
+	if (!priv->pe.port) {
+		return -ENODEV;
+	}
+
+	err = gpio_pin_configure(priv->pe.port, priv->pe.pin,
+				 priv->pe.flags | GPIO_OUTPUT_ACTIVE);
+	if (err) {
+		return err;
+	}
+
+	k_sleep(K_MSEC(50));
+
+	return 0;
+}
+
+static int stm32_sdmmc_pwr_uninit(struct stm32_sdmmc_priv *priv)
+{
+	if (!priv->pe.name) {
+		return 0;
+	}
+
+	gpio_pin_configure(priv->pe.port, priv->pe.pin, GPIO_DISCONNECTED);
+	return 0;
+}
+
+static int disk_stm32_sdmmc_init(struct device *dev)
+{
+	struct stm32_sdmmc_priv *priv = dev->driver_data;
+	int err;
+
+	k_work_init(&priv->work, stm32_sdmmc_cd_handler);
+
+	err = stm32_sdmmc_card_detect_init(priv);
+	if (err) {
+		return err;
+	}
+
+	err = stm32_sdmmc_pwr_init(priv);
+	if (err) {
+		goto err_card_detect;
+	}
+
+	if (stm32_sdmmc_card_present(priv)) {
+		priv->status = DISK_STATUS_UNINIT;
+	} else {
+		priv->status = DISK_STATUS_NOMEDIA;
+	}
+
+	stm32_sdmmc_info.dev = dev;
+	err = disk_access_register(&stm32_sdmmc_info);
+	if (err) {
+		goto err_pwr;
+	}
+	return 0;
+
+err_pwr:
+	stm32_sdmmc_pwr_uninit(priv);
+err_card_detect:
+	stm32_sdmmc_card_detect_uninit(priv);
+	return err;
+}
+
+#if DT_HAS_DRV_INST(0)
```

Comment:
Can you try #if DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(0))?

Reply:
Sorry. Didn't build it myself after rebasing.